Refactor FlowAggregator exporter code

antoninbas opened the pull request:
We currently support 2 different exporters for the FlowAggregator: the
IPFIX exporter and the ClickHouse exporter. As I plan to add a third
one, which will upload flow record to AWS S3, it is good to unify
exporter implementations behind a single interface.

Before this change the IPFIX exporter was deeply embedded within the
main FlowAggregator logic. With this change, it is treated as any other
exporter.

This change also tries to improve thread-safety (in particular for live
updates of exporter configuration options) and adds a few unit tests.

Signed-off-by: Antonin Bas <abas@vmware.com>

wsquan171 commented on pkg/flowaggregator/clickhouseclient/clickhouseclient.go:
```diff
@@ -380,24 +413,32 @@ func (ch *ClickHouseExportProcess) getClickHouseFlowRow(record ipfixentities.Rec
 func (ch *ClickHouseExportProcess) flowRecordPeriodicCommit() {
+	klog.InfoS("Starting ClickHouse exporting process")
+	ctx := context.Background()
+	const flushTimeout = 10 * time.Second
```
do we want it to be pkg scoped rather than func scoped?

wsquan171 reviewed (commented):
thanks Antonin. overall lgtm.

antoninbas commented:
@wsquan171 thanks for the review. I addressed your comments. I also had to make some changes (see last commit) because the new implementation of IPFIX exporter had a bug that was revealed by the e2e tests. I also added some more unit tests.

antoninbas commented:
/test-all

yuntanghsu commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -450,121 +273,42 @@ func (fa *flowAggregator) InitAggregationProcess() error {
-		defer fa.dbExportProcess.Stop()
+	if fa.ipfixExporter != nil {
+		fa.ipfixExporter.Start()
+		defer fa.ipfixExporter.Stop()
```
I think this line should be put inside the flowExportLoop (case <-stopCh). It will panic when we do enable clickhouse/ipfix after we had disabled clickhouse/ipfix. Because the fa.ipfixExporter and fa.clickHouseExporter won't be the original one.

antoninbas replied:
Thanks let me add some unit tests to confirm and I'll fix this.
I think the original code may have suffered from a similar issue. For example with the following sequence:
1) FA is started with ClickHouse enabled
2) CLI is used to disabled ClickHouse
3) FA is stopped (panic because `fa.dbExportProcess.Stop()` is called twice)

That being said, not a super serious bug since it only happens when stopping the FA

antoninbas replied:
@yuntanghsu Please check out the most recent commit. I made the change you suggested and also added a unit test. Let me know if there is a specific sequence of events you want me to test.

yuntanghsu replied:
It looks good now.

yuntanghsu commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -450,121 +273,42 @@ func (fa *flowAggregator) InitAggregationProcess() error {
+	}
+	if fa.clickHouseExporter != nil {
+		fa.clickHouseExporter.Start()
+		defer fa.clickHouseExporter.Stop()
```
ditto

dreamtalen reviewed (dismissed):
Thanks Antonin, LGTM.

yanjunz97 commented on pkg/flowaggregator/exporter/clickhouse.go:
```diff
@@ -0,0 +1,88 @@
+
+func NewClickHouseExporter(opt *options.Options) (*ClickHouseExporter, error) {
+	chInput := buildClickHouseInput(opt)
+	klog.InfoS("Clickhouse configuration", "database", chInput.Database, "databaseURL", chInput.DatabaseURL, "debug", chInput.Debug, "compress", *chInput.Compress, "commitInterval", chInput.CommitInterval)
```
```suggestion
	klog.InfoS("ClickHouse configuration", "database", chInput.Database, "databaseURL", chInput.DatabaseURL, "debug", chInput.Debug, "compress", *chInput.Compress, "commitInterval", chInput.CommitInterval)
```

yanjunz97 commented on pkg/flowaggregator/exporter/clickhouse.go:
```diff
@@ -0,0 +1,88 @@
+	if dsn != e.chExportProcess.GetDsn() {
+		e.chExportProcess.UpdateCH(dsn, connect)
+	}
+	klog.InfoS("New Clickhouse configuration", "database", chInput.Database, "databaseURL", chInput.DatabaseURL, "debug", chInput.Debug, "compress", *chInput.Compress, "commitInterval", chInput.CommitInterval)
```
```suggestion
	klog.InfoS("New ClickHouse configuration", "database", chInput.Database, "databaseURL", chInput.DatabaseURL, "debug", chInput.Debug, "compress", *chInput.Compress, "commitInterval", chInput.CommitInterval)
```

yanjunz97 commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -450,121 +274,59 @@ func (fa *flowAggregator) InitAggregationProcess() error {
+	wg.Add(1)
+	go func() {
+		// We want to make sure that flowExportLoop returns before
+		// returning from this runction. This is because flowExportLoop
```
```suggestion
		// returning from this function. This is because flowExportLoop
```

antoninbas commented:
@heanlan @yanjunz97 Thanks for the reviews, I addressed your comments

antoninbas commented:
/test-all
